SamplesDashboard/SamplesDashboard/Services/NuGetService.cs

```diff
@@ -0,0 +1,102 @@
+    {
+        private readonly SourceRepository _nuGetRepository;
+        private readonly double _cacheLifetime;
+        private readonly IMemoryCache _cache;
+
+        public NuGetService(
+            IConfiguration configuration,
+            IMemoryCache memoryCache)
+        {
+            _cacheLifetime = configuration.GetValue<double>(Constants.CacheLifetime);
+            _cache = memoryCache;
+            _nuGetRepository = NugetRepository.Factory
+                .GetCoreV3("https://api.nuget.org/v3/index.json");
+        }
+
+        public async Task<IEnumerable<NuGetVersion>> GetPackageVersions(string packageName)
+        {
+            using var cache = new SourceCacheContext();
+            var resource = await _nuGetRepository.GetResourceAsync<FindPackageByIdResource>();
+            var versions = await resource.GetAllVersionsAsync(
+                packageName, cache, NullLogger.Instance, CancellationToken.None);
+
+            return versions;
+        }
+
+        public async Task<string> GetLatestVersion(string packageName, string currentVersion)
+        {
+            NuGetVersion latestVersion = null;
+            NuGetVersion latestPreview = null;
+
+            var cacheKey = $"nuget:{packageName}";
+            if (!_cache.TryGetValue(cacheKey, out string latestAndPreview))
+            {
+                var versions = await GetPackageVersions(packageName);
+
+                var nonPreviewVersions = (from version in versions
+                    where !version.IsPrerelease
+                    select version).ToList();
+
+                var previewVersions = (from version in versions
+                    where version.IsPrerelease
+                    select version).ToList();
+
+                latestVersion = nonPreviewVersions.LastOrDefault();
+                latestPreview = previewVersions.LastOrDefault();
+
+                var latestVersionString = latestVersion?.ToString() ?? string.Empty;
+                var latestPreviewString = latestPreview?.ToString() ?? string.Empty;
+
+                // Save the versions into the cache
+                var cacheEntryOptions = new MemoryCacheEntryOptions()
+                    .SetAbsoluteExpiration(TimeSpan.FromSeconds(_cacheLifetime));
+                _cache.Set(cacheKey, $"{latestVersionString};{latestPreviewString}");
+            }
+            else
+            {
+                var versions = latestAndPreview.Split(';');
+                latestVersion = string.IsNullOrEmpty(versions[0]) ? null : new NuGetVersion(versions[0]);
+                latestPreview = versions.Length <= 1 || string.IsNullOrEmpty(versions[1]) ? null : new NuGetVersion(versions[1]);
+            }
+
+            try
+            {
+                var currentVer = new NuGetVersion(currentVersion);
+
+                if (currentVer.IsPrerelease)
+                {
+                    return latestVersion > currentVer ? latestVersion.ToString() : latestPreview.ToString();
```

Comment:
`latestVersion` and `latestPreview` could be null based on line 77 and 78. 
This could also affect line 95.